channels/generic/websocket.py

```diff
@@ -254,3 +256,155 @@ class AsyncJsonWebsocketConsumer(AsyncWebsocketConsumer):
     @classmethod
     async def encode_json(cls, content):
         return json.dumps(content)
+
+
+class AsyncJsonWebsocketDemultiplexer(AsyncJsonWebsocketConsumer):
+
+    """
+    JSON-understanding WebSocket consumer subclass that handles demultiplexing
+    streams using a "stream" key in a top-level dict and the actual payload
+    in a sub-dict called "payload". This lets you run multiple streams over
+    a single WebSocket connection in a standardised way.
+    Incoming messages on streams are dispatched to consumers so you can
+    just tie in consumers the normal way.
+
+    Set a mapping of streams to an application classes in the "applications"
+    keyword.
+    """
+
+    async def receive_json(self, content, **kwargs):
+        """
+        Rout the message to the correct child.
+        """
+
+        # Check the frame looks good
+        if isinstance(content, dict) and "stream" in content and "payload" in content:
+            # Match it to a channel
+            stream = content["stream"]
+            try:
+                child_queue = await self._get_child_queue(stream)
+
+            except ValueError:
+                raise ValueError(
+                    "Invalid multiplexed frame received (stream not mapped)"
+                )
+
+            payload = content["payload"]
+
+            # send it on to the child application
+            child_queue.put_nowait(
+                {
+                    "type": "websocket.receive",
+                    "text": await self.encode_json(payload)
+                }
+            )
+            return
+
+        else:
+            raise ValueError("Invalid multiplexed **frame received (no channel/payload key)")
+
+    async def websocket_send(self, message, child_lookup):
```

Comment:
Can we call `child_lookup` -> `stream_name` everywhere? That would make it clearer.